src/vs/workbench/api/browser/mainThreadChatAgents2.ts

```diff
@@ -321,6 +323,37 @@ export class MainThreadChatAgents2 extends Disposable implements MainThreadChatA
 				this._unresolvedAnchors.get(requestId)?.set(revivedProgress.resolveId, revivedProgress);
 			}
 
+			// Handle ChatPrepareToolInvocationPart: call handleToolStream and generate progress message
+			if (revivedProgress.kind === 'prepareToolInvocation') {
+				const toolData = this._languageModelToolsService.getTool(revivedProgress.toolName);
+				if (toolData) {
+					const streamContext: IToolInvocationStreamContext = {
+						rawInput: revivedProgress.streamData?.partialInput
+					};
+
+					try {
+						const streamResult = await this._languageModelToolsService.handleToolStream(toolData.id, streamContext, CancellationToken.None);
+						let progressContent: IMarkdownString;
+						if (streamResult?.invocationMessage) {
+							progressContent = typeof streamResult.invocationMessage === 'string'
+								? new MarkdownString(streamResult.invocationMessage)
+								: streamResult.invocationMessage;
+						} else {
+							progressContent = new MarkdownString('Invoking tool: ' + toolData.displayName);
+						}
+						// Generate a progress message part from the handleToolStream result
+						const progressMessage: IChatProgressMessage = {
```

Comment:
Not sure I follow why there's a progress message involved here. I'd expect that the partial tool content just gets pushed into the ChatModel and then is rendered similarly to normal tool invocations

Reply:
It seems like you're making a decision about how content should be rendered at this level, when the renderer should decide that